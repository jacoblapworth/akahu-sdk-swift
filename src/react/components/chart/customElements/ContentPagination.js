import React, {Component} from 'react';
// import PropTypes from 'prop-types';
import XUIButton from '../../button/XUIButton';
import XUIIcon from '../../icon/XUIIcon';
import arrowPathData from '@xero/xui-icon/icons/arrow';

class ContentPagination extends Component {
	render() {
		const {current, total, updatePanel, createMessage} = this.props;
		const message = createMessage && createMessage(current, total);

		return (
			<nav
				className="xui-chart--pagination"
				role="navigation"
				aria-label="Pagination">

				<XUIButton
					variant="icon"
					className={`xui-button-icon-large`}
					onClick={() => updatePanel(current - 1)}
					title="Previous page"
					isDisabled={current === 1}>
					<XUIIcon
						path={arrowPathData}
						rotation="90"
					/>
				</XUIButton>

				{message}

				<XUIButton
					variant="icon"
					className={`xui-button-icon-large`}
					onClick={() => updatePanel(current + 1)}
					title="Next Page"
					isDisabled={current === total}>
					<XUIIcon
						path={arrowPathData}
						rotation="270"
					/>
				</XUIButton>

			</nav>
		);
	}
}

export default ContentPagination;
