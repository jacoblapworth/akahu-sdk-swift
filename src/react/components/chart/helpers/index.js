import getGroupPosition from './groupposition';
import createChartPadding from './chartpadding';

const alwaysPositive = value => Math.max(0, value) || 0;
const testIsCloseEnough = (next, previous, threshold = 2) => (
	next > previous - threshold && next < previous + threshold
);

export {getGroupPosition as default, createChartPadding, alwaysPositive, testIsCloseEnough};
