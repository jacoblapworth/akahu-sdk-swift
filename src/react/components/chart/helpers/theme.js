import { VictoryTheme } from 'victory';

const baseFontTheme = {
	fill: 'gray',
	fontFamily: `"Helvetica Neue", Helvetica, Arial, sans-serif`,
	fontSize: 11
};

const axisLineTheme = {
	fill: 'transparent',
	stroke: 'transparent',
	strokeWidth: 0
};

const gridTheme = {
	fill: 'transparent',
	stroke: 'gray',
	strokeWidth: 1
};

const baseChartTheme = { ...VictoryTheme.grayscale };

baseChartTheme.axis.style.axis = {
	...baseChartTheme.axis.style.axis,
	...axisLineTheme
};

baseChartTheme.axis.style.axisLabels = {
	...baseChartTheme.axis.style.axisLabels,
	...baseFontTheme,
};

baseChartTheme.axis.style.grid = {
	...baseChartTheme.axis.style.grid,
	...gridTheme
};

baseChartTheme.axis.style.tickLabels = {
	...baseChartTheme.axis.style.tickLabels,
	...baseFontTheme,
};

const barChartTheme = { ...baseChartTheme };

export { baseChartTheme as default, barChartTheme, baseFontTheme, axisLineTheme };
