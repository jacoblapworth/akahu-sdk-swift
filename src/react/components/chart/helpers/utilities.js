export const alwaysPositive = value => Math.max(0, value) || 0;

export const testIsCloseEnough = (next, previous, threshold = 2) => (
	next > previous - threshold && next < previous + threshold
);

// https://formidable.com/open-source/victory/docs/common-props/#padding
export const createChartPadding = ({xAxisHeight, yAxisWidth}) => ({
	// Allow room for y-axis text to be entered on the axis line but not bleed
	// over the viewbox.
	top: 10,
	// Gap between x-axis line + label height + bottom (with room for scroll bars).
	bottom: 20 + xAxisHeight + 20,
	// Gap between y-axis line + label width.
	left: yAxisWidth + 20,
	// A gap threshold to safegaurd against overflow.
	right: 2
});

export const pause = (items, callback) => {
	const min = items.length || 1 * 10;
	const max = 1000;
	setTimeout(callback, Math.min(min, max));
}
