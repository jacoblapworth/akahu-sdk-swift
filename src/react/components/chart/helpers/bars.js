import cn from 'classnames';
import {createChartPadding} from '../helpers/utilities';
import {createArray} from '../../progressindicator/helpers/utilities';
import {NAME_SPACE, BAR_MIN_WIDTH, BAR_MAX_WIDTH} from '../helpers/constants';
import {createYAxisLabelFormatThunk, createYAxisTickValues} from '../helpers/yaxis';
import AvatarLabel from '../customElements/AvatarLabel';
import StandardLabel from '../customElements/StandardLabel';
import AbbreviationLabel from '../customElements/AbbreviationLabel';

const findMaxTotalBarStacks = ({y}) => y.reduce((acc, value) => acc + value, 0);

const createBarStats = ({barsData, xAxisVisibleItems, viewportWidth, hasPaginationRaw}) => {
	const barsTotal = barsData.length;
	const limitWithLowerThreshold = baseWidth => Math.max(baseWidth, BAR_MIN_WIDTH);
	const limitWithUpperAndLowerThreshold = baseWidth => baseWidth > BAR_MAX_WIDTH
		? BAR_MAX_WIDTH
		: limitWithLowerThreshold(baseWidth)
	const isConstrainedWidth = Boolean(xAxisVisibleItems);

	// What is an "initial" rough estimate of how many bars are going to fit on a
	// single panel.
	const sanitisedWidth = isConstrainedWidth
		// When requesting a "custom" distribution scenario we don't know exactly what
		// the width is going to be just yet so lets take the raw division for now but
		// still making sure that we do not let the width get too small (e.g if the
		// user requested to fit 1000 items on a panel).
		? limitWithLowerThreshold(viewportWidth / xAxisVisibleItems)
		// In a "standard" scenario we limit the bar widths against the static upper
		// and lower thresholds.
		: limitWithUpperAndLowerThreshold(viewportWidth / barsTotal)

	// Now that we have a rough idea of the quantity / width of the bars on a panel
	// we need to make sure all of the dedicated panel content area is filled up.
	//    . - - - - - - .
	//    |    _     ///|
	//    |  _|o|  _ ///| <--- Wasted space that needs to be
	//    | |+|o|_|o|///|      distributed among visible bars.
	//    | |+|o|+|o|///|
	//    ° - - - - - - °
	const wholeBarsPerPanel = Math.floor(viewportWidth / sanitisedWidth);
	const hasMultiplePanels = barsTotal > wholeBarsPerPanel;
	const totalBarsPerPanel = hasPaginationRaw || !hasMultiplePanels
		? Math.min(wholeBarsPerPanel, barsTotal)
		// When there is an overflow with the native scrolling UI we need to make sure
		// to show "half" of the next panels bar as an aesthetic way to convey hidden
		// content to the user.
		: wholeBarsPerPanel + 0.5;
	const barViewports = Math.ceil(barsTotal / totalBarsPerPanel);
	const barWidth = hasMultiplePanels || isConstrainedWidth
		? viewportWidth / totalBarsPerPanel
		: sanitisedWidth;
	const barsWidth = barWidth * barsTotal;

	const barMaxValue = (
		barsData
			.map(findMaxTotalBarStacks)
			.reduce((acc, value) => Math.max(acc, value), 0)
	);

	return {barsWidth, barWidth, barMaxValue, barViewports};
};

const createBarColorStacks = ({barsData, custom, base }) => {
	const maxStacks = barsData.reduce((acc, {y}) => Math.max(acc, y.length), 0);

	return (
		createArray(maxStacks)
			.map((_, index) => custom && custom[index] || base[index % 2])
	);
};

const createActiveBars = (activeBarsRaw, barsData) => (
	Object
		.keys(activeBarsRaw)
		// Does the ID reference exist in the barData array?
		.filter(key => barsData.filter(({id}) => `${id}` === key).length)
		// Is the ID reference relating to an "active" state e.g NOT "false" or "[]".
		.filter((key) => {
			const state = activeBarsRaw[key];
			return Array.isArray(state) ? state.length : Boolean(state);
		})
		// Turn back into an "active" stack ONLY object (only arrays, no booleans).
		.reduce((acc, key) => {
			const stateRaw = activeBarsRaw[key];
			const state = Array.isArray(stateRaw)
				? stateRaw
				: createArray(barsData[key].y.length).map((_, index) => index)

			return {...acc, [key]: state};
		}, {})
);

const enrichParams = (state, props, chartTheme) => {
	const {
		qaHook,
		chartId,
		chartTitle,
		isChartTitleHidden,
		chartDescription,
		chartHeight,
		keyLabel: keyLabelRaw,
		keyTitle,
		barsData: barsDataRaw,
		barColor: barColorRaw,
		activeBars: activeBarsRaw,
		isStacked,
		onBarClick,
		createBarToolTipMessage,
		xAxisVisibleItems,
		yAxisDefaultTopValue,
		createYAxisLabelFormat: createYAxisLabelFormatRaw,
		xAxisType,
		hasPagination: hasPaginationRaw,
		createPaginationMessage,
		paginationNextLabel,
		paginationPreviousLabel,
	} = props;

	const {
		chartWidth,
		yAxisWidth,
		xAxisHeight,
		toolTipPosition,
		toolTipMessage,
		panelCurrent: panelCurrentRaw
	} = state;

	// We support both "plain" and "stacked" bar styles. The difference is that
	// stacks require arrays of data and plain a single value. Rather than create
	// two duplicate components we augment the plain data to mimic a stacked
	// scenario that has only a single stack.
	const keyLabel = keyLabelRaw && (isStacked ? keyLabelRaw : [keyLabelRaw]);
	const barColor = barColorRaw && (isStacked ? barColorRaw : [barColorRaw]);
	const barsData = (
		isStacked
			? barsDataRaw.map(bar => ({...bar, y: bar.y || [0]}))
			: barsDataRaw.map(bar => ({...bar, y: [bar.y || 0]}))
	);

	// Colors...
	const colorStacks = createBarColorStacks({barsData, custom: barColor, base: chartTheme.bar.colorScale});

	// Tooltip...
	const hasToolTip = Boolean(toolTipMessage);

	// Key...
	const hasKey = keyLabel && keyLabel.length;

	// Chart...
	const hasChartTitle = !isChartTitleHidden && chartTitle;
	const hasChartHeader = hasChartTitle || hasKey || hasPagination;
	const isChartNarrow = chartWidth <= 520;
	const chartPadding = createChartPadding({xAxisHeight, yAxisWidth});
	const {top: chartTop, right: chartRight, bottom: chartBottom, left: chartLeft} = chartPadding;

	// Bars...
	const viewportWidth = chartWidth - chartLeft - chartRight;
	const activeBars = activeBarsRaw ? createActiveBars(activeBarsRaw, barsData) : {};
	const {
		barsWidth, barWidth, barMaxValue, barViewports,
	} = createBarStats({barsData, xAxisVisibleItems, viewportWidth, hasPaginationRaw});

	// Panels...
	const panelsTotal = barViewports;
	const panelWidth = viewportWidth;
	// If the user resizes the UI we can get into a situation where the current
	// pagination reference exceeds the available panels.
	const panelCurrent = Math.min(panelCurrentRaw, panelsTotal);

	// Pagination...
	const hasPagination = hasPaginationRaw && panelsTotal > 1;

	// Y-Axis...
	const yAxisHeight = chartHeight - chartTop - chartBottom;
	const {yAxisTickValues, yAxisMaxValue} = createYAxisTickValues({
		yAxisHeight,
		maxValues: [yAxisDefaultTopValue, barMaxValue]
	});
	const createYAxisLabelFormat = createYAxisLabelFormatRaw || createYAxisLabelFormatThunk(yAxisMaxValue);

	// X-Axis...
	const xAxisTickValues = barsData.map(({x}) => x);
	const xAxisLabelOptions = {
		abbreviation: AbbreviationLabel,
		avatar: AvatarLabel,
		standard: StandardLabel,
	};
	const XAxisLabel = xAxisLabelOptions[xAxisType];

	const chartClassName = cn(`${NAME_SPACE}-chart`, {
		[`${NAME_SPACE}-chart-has-pagination`]: hasPagination,
		[`${NAME_SPACE}-chart-has-multiline-header`]: hasPagination && createPaginationMessage && isChartNarrow
	});

	return {
		qaHook,

		// Chart...
		chartId, hasChartHeader, chartTitle, hasChartTitle, chartDescription,
		chartClassName, chartTheme, chartHeight, chartWidth,
		chartPadding, chartTop, chartBottom, chartLeft,

		// Panels...
		panelWidth, panelCurrent, panelsTotal,

		// Bars...
		barsData, barsWidth, barWidth, onBarClick, activeBars,

		// Pagination...
		hasPagination, createPaginationMessage, paginationNextLabel, paginationPreviousLabel,

		// Tooltip...
		toolTipMessage, toolTipPosition, hasToolTip, createBarToolTipMessage,

		// Colors...
		colorStacks,

		// Y-Axis...
		yAxisMaxValue, yAxisHeight, yAxisTickValues, createYAxisLabelFormat,

		// X-Axis...
		xAxisTickValues, XAxisLabel,

		// Key...
		hasKey, keyTitle, keyLabel,
	};
};

export {createBarStats as default, createBarColorStacks, enrichParams, findMaxTotalBarStacks};
