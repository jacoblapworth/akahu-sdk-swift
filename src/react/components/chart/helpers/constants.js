import {ns} from "../../helpers/xuiClassNamespace";

export const NAME_SPACE = `${ns}-chart`;
export const CHART_HEIGHT = 400;
export const CHART_WIDTH = 200;
export const X_AXIS_HEIGHT = 50;
export const Y_AXIS_WIDTH = 50;
export const BAR_MIN_WIDTH = 34;
export const BAR_MAX_WIDTH = 200;
