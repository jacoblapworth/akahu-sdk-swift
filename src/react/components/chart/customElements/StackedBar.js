import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import getTargetPosition from '../helpers/targetposition';
import { NAME_SPACE, BAR_RADIUS, BAR_PADDING_X } from '../helpers/constants';
import { forceValuePositive } from '../helpers/utilities';
import { getMinAndMaxYAxisTickValues } from '../helpers/yaxis';
import {
  createStackTop,
  createStackHeight,
  createInteractionParams,
  testIsAnyStackNegative,
  testIsAnyStackPositive,
  testIsCurrentStackPositive,
  addStackItems,
  forceAddStackItems,
} from '../helpers/bars';

class StackedBar extends PureComponent {
  handleToolTipShow = (event, barData) => {
    const { updateToolTip, createToolTipMessage } = this.props;
    const position = getTargetPosition(event);
    const message = createToolTipMessage(barData);
    updateToolTip(position, message);
  };

  handleToolTipHide = () => {
    const { updateToolTip } = this.props;
    return updateToolTip();
  };

  createBarMask = ({ barLeft, barTop, barWidth, barHeight }) => {
    const {
      datum: { y: stackData },
    } = this.props;
    const isAnyStackNegative = testIsAnyStackNegative(stackData);
    const isAnyStackPositive = testIsAnyStackPositive(stackData);
    const cornerOverrideHeight = Math.min(BAR_RADIUS, barHeight);
    const baseProps = {
      x: barLeft,
      y: barTop,
      width: barWidth,
      height: barHeight,
      fill: 'white',
    };

    return (
      <Fragment>
        <rect {...baseProps} rx={BAR_RADIUS} ry={BAR_RADIUS} />
        {
          // Remove the masks rounded corners on the bars "top" / "bottom" sides
          // that touch the "zero" base line.
        }
        {isAnyStackNegative && !isAnyStackPositive && (
          <rect {...baseProps} height={cornerOverrideHeight} />
        )}
        {!isAnyStackNegative && isAnyStackPositive && (
          <rect
            {...baseProps}
            height={cornerOverrideHeight}
            y={barTop + (barHeight - cornerOverrideHeight)}
          />
        )}
      </Fragment>
    );
  };

  createStackThunk = ({ ratio, barZeroBase, barLeft, barWidth, barStacks }) => (
    barStack,
    stackIndex,
  ) => {
    const {
      isBarStacked,
      onBarClick,
      isToolTipHidden,
      colorStacks,
      colorActive,
      activeBars,
      index: barIndex,
      datum: barData,
    } = this.props;
    const { id: barId } = barData;
    const testIsActive = (activeBars[barId] || []).indexOf(stackIndex) >= 0;
    const stackTop = createStackTop({
      barZeroBase,
      barStacks,
      ratio,
      stackIndex,
    });
    const stackHeight = createStackHeight(barStack * ratio);
    const interactionParams = createInteractionParams(isBarStacked, {
      ...barData,
      barIndex,
      stackIndex,
    });
    const clickProps = onBarClick && {
      onClick: event => onBarClick(event, interactionParams),
      style: { cursor: 'pointer' },
    };
    const toolTipProps = !isToolTipHidden && {
      onMouseEnter: event => this.handleToolTipShow(event, interactionParams),
      onMouseLeave: this.handleToolTipHide,
    };

    return (
      <Fragment key={stackIndex}>
        <rect
          {...{ ...clickProps, ...toolTipProps }}
          fill={colorStacks[stackIndex]}
          height={stackHeight}
          width={barWidth}
          x={barLeft}
          y={stackTop}
        />
        {testIsActive && (
          <rect
            fill={colorActive}
            height={stackHeight}
            style={{ pointerEvents: 'none' }}
            width={barWidth}
            x={barLeft}
            y={stackTop}
          />
        )}
      </Fragment>
    );
  };

  render = () => {
    const {
      chartId,
      barWidth: barWidthRaw,
      yAxisTickValues,
      yAxisHeight,
      padding,

      // Victory...
      datum: barData,
      index: barIndex,
      // Unused Victory references...
      // alignment, data, height, horizontal, index, origin, polar, width, x, x0, y, y0
    } = this.props;

    const { id: barId, y: barStacks } = barData;
    if (!barStacks.length) return null;

    const divider = BAR_PADDING_X;
    const { yAxisMinValue, yAxisMaxValue } = getMinAndMaxYAxisTickValues(yAxisTickValues);
    const yAxisValueSpan = forceValuePositive(yAxisMinValue) + forceValuePositive(yAxisMaxValue);
    const ratio = yAxisHeight / yAxisValueSpan;
    const totalStacksValue = barStacks.reduce(forceAddStackItems, 0);
    const positiveStacksValue = barStacks
      .filter(testIsCurrentStackPositive)
      .reduce(addStackItems, 0);
    const maskId = `${NAME_SPACE}-chart--bar-mask-${chartId}${barId}`;
    const barZeroBase = yAxisMaxValue * ratio + padding.top;
    const barLeft = barWidthRaw * barIndex + divider;
    const barTop = barZeroBase - positiveStacksValue * ratio;
    const barWidth = Math.max(barWidthRaw - divider * 2, 0);
    const barHeight = createStackHeight(totalStacksValue * ratio);
    const createStack = this.createStackThunk({
      ratio,
      barZeroBase,
      barLeft,
      barWidth,
      barStacks,
    });

    // The bar is setup into two main parts.
    //
    // 1. The "stacked" <rect />'s that make up the visual part of the bars. Each
    //    "stack" can have independent click and tooltip interactions applied to it.
    //
    //      My Graph.
    //      ¯¯¯¯¯¯¯¯¯
    //   8 .
    //     |             .----.
    //   6 |             |////|
    //     |     .----.  |////|
    //   4 |     |////|---.///|
    //     | .----.///|   |///|---.
    //   2 | |    |///|   |///|   |
    //     |_ _ _ _ _ _ _ _ _ _ _ _ _ .
    //   0     ()  ()  ()  ()  ()
    //         Vv  Ww  Xx  Yy  Zz
    //
    // 2. The "mask" that rounds the top corners of the entire bar. Bars that
    //    original from the "zero" base line (either above or below) do NOT have
    //    rounded corners from the origination side.

    return (
      <g>
        <defs>
          <mask id={maskId} maskUnits="userSpaceOnUse">
            {this.createBarMask({
              barLeft,
              barTop,
              barWidth,
              barHeight,
            })}
          </mask>
        </defs>

        <g mask={`url(#${maskId})`}>{barStacks.map(createStack)}</g>
      </g>
    );
  };
}

export default StackedBar;

StackedBar.propTypes = {
  chartId: PropTypes.string,
  isBarStacked: PropTypes.bool,
  onBarClick: PropTypes.func,
  isToolTipHidden: PropTypes.bool,
  updateToolTip: PropTypes.func,
  createToolTipMessage: PropTypes.func,
  colorStacks: PropTypes.arrayOf(PropTypes.string),
  colorActive: PropTypes.string,
  activeBars: PropTypes.object,
  datum: PropTypes.object,
  index: PropTypes.number,
  barWidth: PropTypes.number,
  padding: PropTypes.object,
  yAxisTickValues: PropTypes.arrayOf(PropTypes.number),
  yAxisHeight: PropTypes.number,
};
