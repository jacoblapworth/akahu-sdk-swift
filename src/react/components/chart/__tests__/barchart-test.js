import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import { axe, toHaveNoViolations } from 'jest-axe';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import noop from '../../helpers/noop';
import XUIBarChart from '../XUIBarChart';
import StackedBar from '../customElements/StackedBar';
import ContentPagination from '../customElements/ContentPagination';
import { X_AXIS_HEIGHT, Y_AXIS_WIDTH } from '../helpers/constants';
import { createChartPadding } from '../helpers/utilities';
import { enrichParams } from '../helpers/bars';
import { barChartTheme } from '../helpers/theme';

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

const getFirstCallsParams = handlers => handlers.mock.calls[0];

const standardBarsProps = {
  barsData: [
    { id: '123', x: 'Apple', y: 1 },
    { id: '456', x: 'Potato', y: 2 },
    { id: '789', x: 'Carrot', y: 3 },
  ],
};

const standardColorProps = {
  barColor: 'blue',
  barColorActive: 'green',
};

const standardKeyProps = {
  keyTitle: 'Healthy Food',
  keyLabel: 'Fruit',
};

const stackedBarsProps = {
  isBarStacked: true,
  barsData: [
    { id: '123', x: 'Apple', y: [1] },
    { id: '456', x: 'Potato', y: [1, 2] },
    { id: '789', x: 'Carrot', y: [1, 2, 3] },
  ],
};

const defaultChartProps = {
  qaHook: 'myChart',
  chartId: '123',
  chartTitle: 'Chart Test',
  chartDescription: 'Chart Jest test',
  keyTitle: 'Graph key',
  emptyMessage: 'There is no data to display',
  paginationNextTitle: 'Next page',
  paginationPreviousTitle: 'Previous page',
  loadingAriaLabel: 'Loading',
};

const defaultChartPadding = createChartPadding({
  xAxisHeight: X_AXIS_HEIGHT,
  yAxisWidth: Y_AXIS_WIDTH,
});

const defaultBarProps = {
  // XUI...
  chartId: '123',
  yAxisTickValues: [0, 1, 2, 3],
  yAxisHeight: 300,
  colorStacks: ['black'],
  colorActive: 'green',
  activeBars: {},
  barWidth: 100,
  isToolTipHidden: false,
  updateToolTip: noop,
  padding: defaultChartPadding,

  // Victory...
  datum: stackedBarsProps.barsData[0],
  index: 2,
};

const defaultPaginationProps = {
  qaHook: 'myChart',
  current: 2,
  total: 5,
  createMessage: noop,
  updatePanel: noop,
  paginationLabel: 'Pagination',
  paginationNextTitle: 'Next',
  paginationPreviousTitle: 'Previous',
};

const createEnrichedParams = propsRaw => {
  const component = shallow(<XUIBarChart {...propsRaw} />);
  const state = component.dive().state();
  const props = component.props();

  return enrichParams(state, props, barChartTheme);
};

describe('<XUIBarChart />', () => {
  it('should pass accessibility testing', async () => {
    const wrapper = mount(
      <XUIBarChart
        chartId="chartPlainBar"
        chartTitle="Plain Bar Implementation"
        emptyMessage="There is no data to display"
        keyTitle="Graph key"
      />,
    );
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  describe('"Standard" to "Stacked" bar enrichment', () => {
    const propsRaw = {
      ...defaultChartProps,
      ...standardBarsProps,
      ...standardColorProps,
      ...standardKeyProps,
    };
    const { barsData: barsDataRaw, keyLabel: keyLabelRaw, barColor: barColorRaw } = propsRaw;
    const { barsData, keyLabel, barColorStacks } = createEnrichedParams(propsRaw);

    it('Should convert standard bars into stacked variants', () => {
      barsData.forEach(({ y }, index) =>
        expect(y).toEqual(expect.arrayContaining([barsDataRaw[index].y])),
      );
    });

    it('Should convert standard keys into stacked variants', () => {
      expect(keyLabel).toEqual(expect.arrayContaining([keyLabelRaw]));
    });

    it('Should convert standard colors into stacked variants', () => {
      expect(barColorStacks).toEqual(expect.arrayContaining([barColorRaw]));
    });
  });

  describe('Key render conditions', () => {
    const getHasKey = keyLabel =>
      createEnrichedParams({
        ...defaultChartProps,
        ...stackedBarsProps,
        keyLabel,
      }).hasKey;

    it('Should not render a key with empty data', () => {
      const hasKey = getHasKey();

      expect(hasKey).toBeFalsy();
    });

    it('Should not render stacked keys with incomplete data', () => {
      const hasKey = getHasKey(['Fruit']);

      expect(hasKey).toBeFalsy();
    });
  });

  describe('"Standard" bar interaction', () => {
    const handleClick = jest.fn();
    const createMessage = jest.fn();
    const component = mount(
      <svg>
        <StackedBar
          {...defaultBarProps}
          onBarClick={handleClick}
          createToolTipMessage={createMessage}
        />
      </svg>,
    );
    const rect = component.find('g > rect').at(0);
    const barObject = {
      id: '123',
      x: 'Apple',
      y: 1,
      barIndex: 2,
    };

    it('Should return the current "Standard" bar data set on click', () => {
      rect.simulate('click');
      const [, params] = getFirstCallsParams(handleClick);

      expect(params).toMatchObject(barObject);
    });

    it('Should return the current "Standard" bar data set on mouse enter', () => {
      rect.simulate('mouseEnter');
      const [params] = getFirstCallsParams(createMessage);

      expect(params).toMatchObject(barObject);
    });
  });

  describe('"Stacked" bar interaction', () => {
    const handleClick = jest.fn();
    const createMessage = jest.fn();
    const component = mount(
      <svg>
        <StackedBar
          {...defaultBarProps}
          isBarStacked
          onBarClick={handleClick}
          createToolTipMessage={createMessage}
        />
      </svg>,
    );
    const rect = component.find('g > rect').at(0);
    const barObject = {
      id: '123',
      x: 'Apple',
      y: [1],
      barIndex: 2,
      stackIndex: 0,
    };

    it('Should pass on the current "Stacked" bar data set on click', () => {
      rect.simulate('click');
      const [, params] = getFirstCallsParams(handleClick);

      expect(params).toMatchObject(barObject);
    });

    it('Should pass on the current "Stacked" bar data set on mouse enter', () => {
      rect.simulate('mouseEnter');
      const [params] = getFirstCallsParams(createMessage);

      expect(params).toMatchObject(barObject);
    });
  });

  describe('Pagination interaction', () => {
    it('Should pass on the current panel relationship when targeting the "previous" panel', () => {
      const createMessage = jest.fn();
      mount(<ContentPagination {...defaultPaginationProps} createMessage={createMessage} />);
      const params = getFirstCallsParams(createMessage);

      expect(params).toEqual(expect.arrayContaining([2, 5]));
    });

    it('Should disable the "previous" button when the current panel is at its "minimum" value', () => {
      const component = renderer.create(
        <ContentPagination {...defaultPaginationProps} current={1} />,
      );
      expect(component).toMatchSnapshot();
    });

    it('Should disable the "next" button when the current panel is at its "maximum" value', () => {
      const component = renderer.create(
        <ContentPagination {...defaultPaginationProps} current={defaultPaginationProps.total} />,
      );
      expect(component).toMatchSnapshot();
    });
  });
});
