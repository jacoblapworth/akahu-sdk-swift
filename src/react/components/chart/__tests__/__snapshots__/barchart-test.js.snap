// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUIBarChart /> Pagination interaction Should disable the "next" button when the current panel is at its "maximum" value 1`] = `
<nav
  aria-label="Pagination"
  className="xui-chart--pagination"
  data-automationid="myChart--pagination"
>
  <button
    className="xui-button xui-button-icon xui-button-icon-medium"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
    title="Previous"
    type="button"
  >
    <svg
      className="xui-icon xui-icon-rotate-90"
      focusable="false"
      height={7}
      viewBox="0 0 11 7"
      width={11}
    >
      <path
        d="M5.5 4.2L9.625 0 11 1.4 5.5 7 0 1.4 1.375 0z"
        role="presentation"
      />
    </svg>
  </button>
  <button
    className="xui-button xui-button-icon xui-button-icon-medium xui-button-is-disabled"
    disabled={true}
    onClick={[Function]}
    onKeyDown={null}
    tabIndex={-1}
    title="Next"
    type="button"
  >
    <svg
      className="xui-icon xui-icon-rotate-270"
      focusable="false"
      height={7}
      viewBox="0 0 11 7"
      width={11}
    >
      <path
        d="M5.5 4.2L9.625 0 11 1.4 5.5 7 0 1.4 1.375 0z"
        role="presentation"
      />
    </svg>
  </button>
</nav>
`;

exports[`<XUIBarChart /> Pagination interaction Should disable the "previous" button when the current panel is at its "minimum" value 1`] = `
<nav
  aria-label="Pagination"
  className="xui-chart--pagination"
  data-automationid="myChart--pagination"
>
  <button
    className="xui-button xui-button-icon xui-button-icon-medium xui-button-is-disabled"
    disabled={true}
    onClick={[Function]}
    onKeyDown={null}
    tabIndex={-1}
    title="Previous"
    type="button"
  >
    <svg
      className="xui-icon xui-icon-rotate-90"
      focusable="false"
      height={7}
      viewBox="0 0 11 7"
      width={11}
    >
      <path
        d="M5.5 4.2L9.625 0 11 1.4 5.5 7 0 1.4 1.375 0z"
        role="presentation"
      />
    </svg>
  </button>
  <button
    className="xui-button xui-button-icon xui-button-icon-medium"
    disabled={false}
    onClick={[Function]}
    tabIndex={0}
    title="Next"
    type="button"
  >
    <svg
      className="xui-icon xui-icon-rotate-270"
      focusable="false"
      height={7}
      viewBox="0 0 11 7"
      width={11}
    >
      <path
        d="M5.5 4.2L9.625 0 11 1.4 5.5 7 0 1.4 1.375 0z"
        role="presentation"
      />
    </svg>
  </button>
</nav>
`;
